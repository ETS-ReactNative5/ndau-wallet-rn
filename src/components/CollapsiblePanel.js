import React, { Component } from 'react';
import {
  StyleSheet,
  Text,
  View,
  ImageBackground,
  TouchableHighlight,
  Animated,
  Image,
  PixelRatio
} from 'react-native';
import NdauNodeAPIHelper from '../helpers/NdauNodeAPIHelper';
import {
  widthPercentageToDP as wp,
  heightPercentageToDP as hp
} from 'react-native-responsive-screen';
import cssStyles from '../css/styles';

class CollapsiblePanel extends Component {
  constructor(props) {
    super(props);

    this.cardBackgrounds = [
      require('../../img/green-card.png'),
      require('../../img/light-blue-card.png'),
      require('../../img/dark-blue-card.png'),
      require('../../img/green-locked-card.png'),
      require('../../img/light-blue-locked-card.png'),
      require('../../img/dark-blue-locked-card.png'),
      require('../../img/grey-notified-card.png')
    ];

    this.state = {
      expanded: true,
      animation: new Animated.Value(),
      maxHeight: 0,
      minHeight: 0
    };
  }

  toggle() {
    this.setState({
      expanded: !this.state.expanded
    });

    let initialValue = this.state.expanded
        ? this.state.maxHeight + this.state.minHeight
        : this.state.minHeight,
      finalValue = this.state.expanded
        ? this.state.minHeight
        : this.state.maxHeight + this.state.minHeight;

    this.state.animation.setValue(initialValue);
    Animated.spring(this.state.animation, {
      toValue: finalValue
    }).start();
  }

  setMaxHeight = (event) => {
    let adjustment = hp('1.2%');
    if (event.nativeEvent.layout.height > this.state.maxHeight) {
      console.debug(`setting maxHeight for first time: ${event.nativeEvent.layout.height}`);
      this.setState({
        maxHeight: event.nativeEvent.layout.height + adjustment
      });
    }
  };

  setMinHeight = (event) => {
    this.setState({
      minHeight: event.nativeEvent.layout.height
    });
  };

  render() {
    const lockAdder = this.props.lockAdder ? this.props.lockAdder : 0;
    return (
      <Animated.View style={[ styles.container, { height: this.state.animation } ]}>
        <ImageBackground
          source={this.cardBackgrounds[this.props.onNotice ? 6 : this.props.index % 3 + lockAdder]}
          style={{ width: '100%' }}
        >
          <TouchableHighlight onPress={this.toggle.bind(this)} underlayColor="transparent">
            <View style={styles.titleContainer} onLayout={this.setMinHeight}>
              <Text style={styles.titleLeft}>{this.props.title}</Text>
              {this.props.account ? (
                <View
                  style={{ flexDirection: 'row', alignItems: 'center', justifyContent: 'center' }}
                >
                  <Image
                    style={{
                      width: wp('3%'),
                      height: hp('2%'),
                      marginRight: wp('.1%')
                    }}
                    resizeMode="contain"
                    source={require('../../img/ndau-icon-white.png')}
                  />
                  <Text style={styles.titleRight}>
                    {NdauNodeAPIHelper.accountNdauAmount(this.props.account)}
                  </Text>
                </View>
              ) : null}
            </View>
          </TouchableHighlight>
          <View style={styles.border} />
          <View style={styles.body} onLayout={this.setMaxHeight}>
            {this.props.children}
          </View>
        </ImageBackground>
      </Animated.View>
    );
  }
}

var styles = StyleSheet.create({
  container: {
    backgroundColor: 'transparent',
    overflow: 'hidden',
    marginBottom: hp('.5%'),
    borderRadius: 6,
    borderWidth: 0.5
  },
  titleContainer: {
    flexDirection: 'row'
  },
  titleLeft: {
    flex: 1,
    marginTop: hp('.5%'),
    marginLeft: wp('1%'),
    marginRight: wp('1%'),
    marginBottom: hp('.5%'),
    color: '#fff',
    fontSize: 18,
    fontFamily: 'TitilliumWeb-Bold',
    textAlign: 'left'
  },
  titleRight: {
    flexDirection: 'row',
    marginTop: hp('.5%'),
    marginLeft: wp('1%'),
    marginRight: wp('1%'),
    marginBottom: hp('.5%'),
    color: '#fff',
    fontSize: 18,
    fontFamily: 'TitilliumWeb-Bold',
    textAlign: 'right'
  },
  body: {
    padding: hp('1%'),
    paddingTop: 0
  },
  border: {
    borderBottomColor: 'white',
    borderBottomWidth: 1,
    marginLeft: wp('1%'),
    marginRight: wp('1%'),
    marginBottom: hp('1%')
  }
});

export default CollapsiblePanel;
