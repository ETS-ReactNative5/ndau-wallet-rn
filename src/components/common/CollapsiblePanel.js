import React, { Component } from 'react'
import { Text, View, TouchableHighlight, Animated } from 'react-native'
import {
  widthPercentageToDP as wp,
  heightPercentageToDP as hp
} from 'react-native-responsive-screen'
import styles from './styles'

class CollapsiblePanel extends Component {
  constructor (props) {
    super(props)

    this.state = {
      expanded: false,
      animation: new Animated.Value(),
      maxHeight: 0,
      minHeight: 0,
      bodyData: null
    }
  }

  toggle = () => {
    this.setState({
      expanded: !this.state.expanded,
      bodyData: this.props.children
    })

    let initialValue = this.state.expanded
      ? this.state.maxHeight + this.state.minHeight
      : this.state.minHeight

    let finalValue = this.state.expanded
      ? this.state.minHeight
      : this.state.maxHeight + this.state.minHeight

    if (!this.state.animation) {
      this.setState({ animation: new Animated.Value() }, () => {
        this.state.animation.setValue(initialValue)
        Animated.spring(this.state.animation, {
          toValue: finalValue
        }).start()
      })
    } else {
      this.state.animation.setValue(initialValue)
      Animated.spring(this.state.animation, {
        toValue: finalValue
      }).start()
    }
  }

  setMaxHeight = event => {
    let adjustment = hp('1%')
    if (event.nativeEvent.layout.height > this.state.maxHeight) {
      this.setState({
        maxHeight: event.nativeEvent.layout.height + adjustment
      })
    }
  }

  setMinHeight = event => {
    if (!this.state.minHeight) {
      this.setState({
        minHeight: event.nativeEvent.layout.height
      })
    }
  }

  componentDidMount () {
    if (!this.state.expanded) {
      this.toggle()

      this.setState(state => ({
        animation: state.animation.setValue(state.minHeight || 10)
      }))
    }
  }

  render () {
    return (
      <Animated.View
        style={[
          styles.collapsiblePanelContainer,
          { height: this.state.animation }
        ]}
      >
        <TouchableHighlight
          onPress={() => this.toggle()}
          underlayColor='transparent'
        >
          <View
            style={styles.collapsiblePanelTitleContainer}
            onLayout={this.setMinHeight}
          >
            <View>
              <Text style={styles.collapsiblePanelTitleLeft}>
                {this.props.title}
              </Text>
            </View>
            <View>
              {this.props.titleRight !== undefined ? (
                <Text style={styles.collapsiblePanelTitleRight}>
                  {this.props.titleRight}
                </Text>
              ) : null}
            </View>
          </View>
        </TouchableHighlight>
        <View style={styles.collapsiblePanelBorder} />
        <View style={styles.collapsiblePanelBody} onLayout={this.setMaxHeight}>
          {this.state.bodyData}
        </View>
      </Animated.View>
    )
  }
}

export default CollapsiblePanel
