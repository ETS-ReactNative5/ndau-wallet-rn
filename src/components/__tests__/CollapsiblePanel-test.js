import { mount } from 'enzyme';
import React from 'react';
import { StyleSheet } from 'react-native';
import CollapsiblePanel from '../CollapsiblePanel';

import renderer from 'react-test-renderer';

describe('testing CollapsiblePanel...', () => {
  let styles = StyleSheet.create({
    wizardText: {
      color: '#ffffff',
      fontSize: 20
    }
  });
  const navigation = {
    navigate: () => {}
  };
  const account = {
    Balance: 200.0,
    TransferKeys: null,
    RewardsTarget: 'tnai24puxki6s4zqyy7ebizgcviw2ui9z9x98pmah5n3ynmz',
    IncomingRewardsFrom: null,
    DelegationNode: null,
    Lock: {
      NoticePeriod: 123491567000001,
      UnlocksOn: 1585886400000
    },
    Stake: null,
    LastEAIUpdate: 589991567000000,
    LastWAAUpdate: 589991567000000,
    WeightedAverageAge: 0,
    Sequence: 0,
    Settlements: null,
    SettlementSettings: { Period: 0, ChangesAt: null, Next: null },
    ValidationScript: null
  };

  it('renders correctly', () => {
    const tree = renderer
      .create(<CollapsiblePanel parentStyles={styles} navigation={navigation} account={account} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
