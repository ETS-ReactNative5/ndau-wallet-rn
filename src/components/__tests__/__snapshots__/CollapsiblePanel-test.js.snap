// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing CollapsiblePanel... renders correctly 1`] = `
<View
  style={
    Object {
      "backgroundColor": "transparent",
      "borderRadius": 6,
      "borderWidth": 0.5,
      "height": NaN,
      "marginBottom": 6.5,
      "overflow": "hidden",
    }
  }
>
  <View
    style={
      Object {
        "width": "100%",
      }
    }
  >
    <Image
      style={
        Array [
          Object {
            "bottom": 0,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
          },
          Object {
            "height": undefined,
            "width": "100%",
          },
          undefined,
        ]
      }
    />
    <View
      accessible={true}
      isTVSelectable={true}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={null}
    >
      <View
        onLayout={[Function]}
        style={
          Object {
            "flexDirection": "row",
          }
        }
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Object {
              "color": "#fff",
              "flex": 1,
              "fontFamily": "TitilliumWeb-Bold",
              "fontSize": 18,
              "marginBottom": 6.5,
              "marginLeft": 7.5,
              "marginRight": 7.5,
              "marginTop": 6.5,
              "textAlign": "left",
            }
          }
        />
        <View
          style={
            Object {
              "flexDirection": "row",
            }
          }
        >
          <Image
            resizeMode="contain"
            source={
              Object {
                "testUri": "../../../img/ndau-icon-white.png",
              }
            }
            style={
              Object {
                "height": 26.5,
                "marginRight": 1,
                "marginTop": 24.5,
                "width": 22.5,
              }
            }
          />
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            style={
              Object {
                "color": "#fff",
                "flexDirection": "row",
                "fontFamily": "TitilliumWeb-Bold",
                "fontSize": 18,
                "marginBottom": 6.5,
                "marginLeft": 7.5,
                "marginRight": 7.5,
                "marginTop": 6.5,
                "textAlign": "right",
              }
            }
          >
            NaN
          </Text>
        </View>
      </View>
    </View>
    <View
      style={
        Object {
          "borderBottomColor": "white",
          "borderBottomWidth": 1,
          "marginBottom": 13.5,
          "marginLeft": 7.5,
          "marginRight": 7.5,
        }
      }
    />
    <View
      onLayout={[Function]}
      style={
        Object {
          "padding": 13.5,
          "paddingTop": 0,
        }
      }
    />
  </View>
</View>
`;
